src/plugins/vis_type_vega/public/data_model/search_api.ts

```diff
@@ -54,31 +58,75 @@ export class SearchAPI {
   ) {}
 
-  search(searchRequests: SearchRequest[]) {
+  async search(searchRequests: SearchRequest[]) {
     const { search } = this.dependencies.search;
     const requestResponders: any = {};
 
     return combineLatest(
-      searchRequests.map((request) => {
-        const requestId = request.name;
-        const params = getSearchParamsFromRequest(request, {
-          getConfig: this.dependencies.uiSettings.get.bind(this.dependencies.uiSettings),
-        });
-
-        if (this.inspectorAdapters) {
-          requestResponders[requestId] = this.inspectorAdapters.requests.start(requestId, request);
-          requestResponders[requestId].json(params.body);
-        }
-
-        return search({ params }, { abortSignal: this.abortSignal }).pipe(
-          tap((data) => this.inspectSearchResult(data, requestResponders[requestId])),
-          map((data) => ({
-            name: requestId,
-            rawResponse: data.rawResponse,
-          }))
-        );
-      })
+      await Promise.all(
+        searchRequests.map(async (request) => {
+          const requestId = request.name;
+          const dataSourceId = !!request.data_source_name
+            ? await this.findDataSourceIdbyName(request.data_source_name)
+            : undefined;
+
+          const params = getSearchParamsFromRequest(request, {
+            getConfig: this.dependencies.uiSettings.get.bind(this.dependencies.uiSettings),
+          });
+
+          if (this.inspectorAdapters) {
+            requestResponders[requestId] = this.inspectorAdapters.requests.start(
+              requestId,
+              request
+            );
+            requestResponders[requestId].json(params.body);
+          }
+
+          const searchApiParams =
+            dataSourceId && this.dependencies.dataSourceEnabled
+              ? { params, dataSourceId }
+              : { params };
+
+          return search(searchApiParams, { abortSignal: this.abortSignal }).pipe(
+            tap((data) => this.inspectSearchResult(data, requestResponders[requestId])),
+            map((data) => ({
+              name: requestId,
+              rawResponse: data.rawResponse,
+            }))
+          );
+        })
+      )
+    );
+  }
+
+  async findDataSourceIdbyName(dataSourceName: string) {
+    if (!this.dependencies.dataSourceEnabled) {
+      throw new Error('data_source_name cannot be used because data_source.enabled is false');
+    }
+    const dataSources = await this.dataSourceFindQuery(dataSourceName);
+
+    // In the case that data_source_name is a prefix of another name, match exact data_source_name
+    const possibleDataSourceIds = dataSources.savedObjects.filter(
+      (obj) => obj.attributes.title === dataSourceName
     );
+
+    if (possibleDataSourceIds.length !== 1) {
+      throw new Error(
+        `Expected exactly 1 result for data_source_name "${dataSourceName}" but got ${possibleDataSourceIds.length} results`
+      );
+    }
```

Comment:
This logic is to check no duplicate datasource names between all datasources.

But we set the paging size 10 here: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5975/files#diff-f1c265abe9e3b3c11a64d4110cf3e6a010dfe369cf6eed9a20ea532763489fbdR125

In that case, will the duplication check only applies to first 10 datasoruces?

Reply:
This comment is not a blocker of this PR.
I don't think we are able to create datasource with duplication name.